kotlin/src/main/kotlin/net/jqwik/kotlin/api/CombinatorDsl.kt

```diff
@@ -0,0 +1,162 @@
+package net.jqwik.kotlin.api
+
+import net.jqwik.api.Arbitrary
+import net.jqwik.api.Combinators
+import net.jqwik.api.Combinators.ListCombinator
+import org.apiguardian.api.API
+import java.util.concurrent.atomic.AtomicBoolean
+import kotlin.properties.ReadOnlyProperty
+import kotlin.reflect.KProperty
+
+/**
+ * Combine arbitraries using the combinator DSL:
+ *
+ * ```kotlin
+ * combine {
+ *     val first by Arbitraries.strings()
+ *     val second by Arbitraries.strings()
+ *
+ *     filter { first.isNotEmpty() }
+ *     filter { first != second }
+ *
+ *     createAs {
+ *         "first: $first, second: $second"
+ *     }
+ * }
+ * ```
+ *
+ * @return new Arbitrary instance
+ */
+@API(status = API.Status.EXPERIMENTAL, since = "1.8.0")
+fun <R> combine(
+    combinator: CombinatorScope.() -> Combined<R>
+): Arbitrary<R> {
+    val bindings = ValueBindings()
+    val combined = CombinatorScope(bindings).combinator()
+    val listCombinator = combined.createListCombinator(bindings)
+
+    return listCombinator.`as` { values ->
+        bindings.withValues(values) {
+            combined.creator()
+        }
+    } as Arbitrary<R>
+}
+
+/**
+ * Flat-combine arbitraries using the combinator DSL:
+ *
+ * ```kotlin
+ * flatCombine {
+ *     val first by Arbitraries.strings()
+ *     val second by Arbitraries.strings()
+ *
+ *     filter { first.isNotEmpty() }
+ *     filter { first != second }
+ *
+ *     createAs {
+ *         Arbitraries.just("first: $first, second: $second")
+ *     }
+ * }
+ * ```
+ *
+ * @return new Arbitrary instance
+ */
+@API(status = API.Status.EXPERIMENTAL, since = "1.8.0")
+fun <R> flatCombine(
+    combinator: CombinatorScope.() -> Combined<Arbitrary<R>>
+): Arbitrary<R> {
```

Comment:
Have you considered using a single `combine` block, and use `combineAs` and `flatCombineAs` factories?
It does not sound right to me to have several top-level `combine` and `flatCombine` builders.